File: JANUARY/13-01-2026/3453.dart
class Solution {
  double separateSquares(List<List<int>> squares) {
    final events = <List<double>>[];

    for (final s in squares) {
      final y = s[1].toDouble();
      final l = s[2].toDouble();
      events.add([y, l]);
      events.add([y + l, -l]);
    }

    events.sort((a, b) => a[0].compareTo(b[0]));

    double curWidth = 0.0;
    double prevY = events[0][0];
    double totalArea = 0.0;
    final segments = <List<double>>[];

    for (final e in events) {
      final y = e[0];
      final delta = e[1];
      if (y > prevY && curWidth > 0) {
        final area = (y - prevY) * curWidth;
        segments.add([prevY, y, curWidth, area]);
        totalArea += area;
      }
      curWidth += delta;
      prevY = y;
    }

    final target = totalArea / 2.0;
    double acc = 0.0;

    for (final seg in segments) {
      final y1 = seg[0];
      final w = seg[2];
      final area = seg[3];
      if (acc + area >= target) {
        return y1 + (target - acc) / w;
      }
      acc += area;
    }

    return 0.0;
  }
}

File: JANUARY/13-01-2026/3453.rb
def separate_squares(squares)
  events = []
  squares.each do |_, y, l|
    events << [y.to_f, l.to_f]
    events << [(y + l).to_f, -l.to_f]
  end

  events.sort_by!(&:first)

  segments = []
  cur_width = 0.0
  prev_y = events[0][0]
  total_area = 0.0

  events.each do |y, delta|
    if y > prev_y && cur_width > 0
      area = (y - prev_y) * cur_width
      segments << [prev_y, y, cur_width, area]
      total_area += area
    end
    cur_width += delta
    prev_y = y
  end

  target = total_area / 2.0
  acc = 0.0

  segments.each do |y1, y2, w, area|
    if acc + area >= target
      return y1 + (target - acc) / w
    end
    acc += area
  end

  0.0
end